### PR TITLE
managedfields: A few improvements which will make testing easier

### PR DESCRIPTION
**Pull request by apelisse**
Allows creating a typeconverter from a client (i.e. by taking the data of the client and formatting it so that one can create a type converter).

I'm trying to see if that's what we want, possible alternatives:
- We create that from the Root object
- We just make a function to merge the specs
- ...?

This needs tests.

#### What type of PR is this?
/kind cleanup

#### Which issue(s) this PR fixes:
Nothing

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

/assign @alexzielenski

### CODE REVIEW

**Review by arimallick — APPROVED** [review]

**Comment by apelisse** [comment]
/hold

This is still a WIP, not sure if we want that.

**Comment by apelisse** [comment]
In tests specifically, we usually need to create a fieldmanager from a test client (either the embedded or the one that loads from files). This creates the testConverter and then you can use it to create a specific field manager for test. It could be even more convenient but I'm not quite sure what the right thing to do is here.

File: staging/src/k8s.io/apimachinery/pkg/util/managedfields/internal/versioncheck.go [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+type versionCheckManager struct {
+	fieldManager Manager
+	gvk          schema.GroupVersionKind
+}
+
+var _ Manager = &versionCheckManager{}
+
+// NewVersionCheckManager creates a manager that makes sure that the
+// applied object is in the proper version.
+func NewVersionCheckManager(fieldManager Manager, gvk schema.GroupVersionKind) Manager {
```
Comment (alexzielenski) [inline staging/src/k8s.io/apimachinery/pkg/util/managedfields/internal/versioncheck.go:36]: how does version check manager help testability. seems like an orthogonal change

**Review by alexzielenski — APPROVED** [review]
/lgtm

version check refactors seems orthogonal but looks OK. `NewFakeFieldManager` and `NewTypeConverter` lgtm

File: staging/src/k8s.io/apimachinery/pkg/util/managedfields/internal/versioncheck.go (hunk 1 above)
Comment (apelisse) [reply staging/src/k8s.io/apimachinery/pkg/util/managedfields/internal/versioncheck.go:36]: Yes, so, in tests, it's very unlikely that we will have one FieldManager per individual type like we do in the apiserver. And at least the SkipProbababilityManager (for now, but I expect more are going to have the same problem) was struggling to create a new object because of that. Hiding the GVK makes some things easier.

File: staging/src/k8s.io/client-go/openapi/typeconverter.go [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+import (
+	"encoding/json"
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/util/managedfields"
+	"k8s.io/kube-openapi/pkg/spec3"
+	"k8s.io/kube-openapi/pkg/validation/spec"
+)
+
+func NewTypeConverter(client Client, preserveUnknownFields bool) (managedfields.TypeConverter, error) {
```
Comment (seans3) [inline staging/src/k8s.io/client-go/openapi/typeconverter.go:28]: Nit: Should there be documentation for this function?

**Review by seans3 — APPROVED** [review]
Nit on documentation. You can fix it next time if you agree.

/lgtm
/approve

**Comment by apelisse** [comment]
/hold cancel

I think we're good with this.
